core/src/main/java/org/infinispan/protostream/impl/StringUtil.java

```diff
@@ -0,0 +1,155 @@
+package org.infinispan.protostream.impl;
+
+import static java.lang.invoke.MethodType.methodType;
+
+import java.lang.invoke.CallSite;
+import java.lang.invoke.LambdaMetafactory;
+import java.lang.invoke.MethodHandle;
+import java.lang.invoke.MethodHandles;
+import java.lang.invoke.MethodType;
+import java.lang.reflect.Field;
+import java.lang.reflect.Method;
+import java.nio.charset.StandardCharsets;
+import java.util.function.Function;
+import java.util.function.ToIntFunction;
+
+import sun.misc.Unsafe;
+
+class StringUtil {
+
+   interface Utf8Helper {
+      boolean providesLatin1Bytes();
+      byte[] getBytes(String s);
+   }
+
+   private static final Log log = Log.LogFactory.getLog(ProtoStreamReaderImpl.class);
+   private static final Byte LATIN1 = (byte) 0;
+
+   static volatile Utf8Helper INSTANCE;
```

Comment:
I don't get why you cannot just have a `static final VarHandle CODER_FIELD` to use.
Since the JVM trust static final fields to treat them as constants, this is necessary to get the proper expected performance...